File: Submissions/120_Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) { 
        int n = triangle.size();
        if(n == 1) return triangle[0][0];
        
        vector<int> dp(n, 0);
        dp[0] = triangle[0][0] + triangle[1][0];
        dp[1] = triangle[0][0] + triangle[1][1];
        
        for(int i=2; i<n; ++i){
            int m = triangle[i].size();
            dp[m-1] = dp[m-2] + triangle[i][m-1];
            for(int j=m-2; j>0; --j){
                dp[j] = min(dp[j-1], dp[j]) + triangle[i][j];
            }
            dp[0] = dp[0] + triangle[i][0];
        }
        
        return *min_element(dp.begin(), dp.end()); 
    }
};

File: Kotlin/120_Triangle.kt
class Solution {
    fun minimumTotal(triangle: List<List<Int>>): Int {
        val n = triangle.size
        if(n == 1) return triangle[0][0]
        
        var dp = MutableList<Int>(triangle.size) { 0 }
        dp[0] = triangle[0][0] + triangle[1][0]
        dp[1] = triangle[0][0] + triangle[1][1]
        
        for(i in 2 until n){
            val m = triangle[i].size
            dp[m - 1] = dp[m - 2] + triangle[i][m - 1]
            for(j in m - 2 downTo 1){
                dp[j] = minOf(dp[j - 1], dp[j]) + triangle[i][j]
            }
            dp[0] = dp[0] + triangle[i][0]
        }
        
        return dp.min()!!
    }
}